### ISSUE TITLE
RealmChangeListener doesn't get notification on transaction complete

### ISSUE THREAD

**Issue by VasyaFromRussia** [realm/realm-java#1859]
I've updated realm to version 0.85.1 in my project.

I've got such code executing in background looper thread: 
```
Realm realm = Realm.getDefaultInstance();
try {
     realm.executeTransaction(trRealm -> trRealm.copyToRealmOrUpdate(message));
} finally {
     realm.close();
}
```

And such code executing in main thread:

```
Realm realm = Realm.getDefaultInstance();
RealmResults<ServerChatMessage> messages = realm.where(ServerChatMessage.class).findAll();
realm.addChangeListener(() -> Logger.d("realm", "realm updated" + messages.size()));
```

This listener never fires. But if I remove 'messages' variable reference (messages.size()), it fires on each realm's update. Where is mistake?

**Comment by nhachicha** [comment]
Hi @VasyaFromRussia 

We used to store listeners as WeakReferences, this will make your anonymous class eligible for Garbage 
Collection. This is why when you add `messages` to it, it will hold a reference the enclosing class containing `messages` (Btw, this is also the behaviour of lambda in Java 8, if inside your anonymous class/lambda you don't reference any variable from the enclosing class it will not hold a reference to it)

The fix is already in master (you can try it with snapshot), this will be released soon in  `0.95.2`. 

Cheers,

**Comment by VasyaFromRussia** [comment]
It works, thanks a lot!
